### PR TITLE
Complex SV intervals support

### PR DESCRIPTION
**Pull request by mwalker174**
- Complex (CPX) SV records no longer use `END2`/`CHR2` INFO fields and instead are defined by `CPX_INTERVALS` in addition to normal coordinates.
- Adds new class `SVCallRecord.ComplexEventInterval` for storing complex event intervals, which are associated with a particular SV type.
- SV clustering now takes into account CPX subtype and intervals, requiring intervals to match on type and interval similarity as defined by the clustering parameters (e.g. minimum reciprocal overlap).
- Improved reciprocal translocation (CTX) record handling. In particular, these records are now treated similarly to BNDs for clustering purposes.
- Added unit tests for new CPX cases.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java [hunk 1]
```diff
@@ -222,12 +274,31 @@ private static int getLength(final SVCallRecord record, final int missingInserti
...
+    /**
+     * Returns length to be used for complex variant matching
+     */
+    private static int getLengthComplex(final SVCallRecord record, final int missingInsertionLength) {
+        if (!record.isIntrachromosomal() || record.getPositionA() == record.getPositionB()) {
+            // Insertion types use the sum of cpx intervals, or the "missing" length if there are none
+            if (record.getComplexEventIntervals().isEmpty()) {
+                return missingInsertionLength;
+            } else {
+                return record.getComplexEventIntervals().stream().mapToInt(SVCallRecord.ComplexEventInterval::getLengthOnReference).sum();
```
Comment (epiercehoffman) [inline src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:294]: I think there are some issues with this length calculation. a) The intervals in CPX_INTERVALS can overlap - ie. in a dDUP type event, there can be an INV interval that is equal to the DUP interval. b) INS_iDEL events (if they should be included here - see above comment) do not contain an interval in CPX_INTERVALS that represents the source sequence, so you would only have the length of the deleted sequence, which may or may not be the desired behavior here.

I believe you could address (a) for insertion-type CPX events if you disregard any INV intervals. 

Addressing (b) depends on the desired behavior - you could look in SOURCE for the inserted sequence, but SOURCE is sometimes removed from the VCF depending on when this code is applied. But a single length measure is hard to make sense of if you're adding together the length of an inserted sequence and a deletion at the sink (ie. in an INS_iDEL or dDUP_iDEL).

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java [hunk 2]
```diff
@@ -222,12 +274,31 @@ private static int getLength(final SVCallRecord record, final int missingInserti
...
             // TODO lengths less than 1 shouldn't be valid
             return Math.max(record.getLength() == null ? 1 : record.getLength(), 1);
         }
     }
 
+    /**
+     * Returns length to be used for complex variant matching
+     */
+    private static int getLengthComplex(final SVCallRecord record, final int missingInsertionLength) {
+        if (!record.isIntrachromosomal() || record.getPositionA() == record.getPositionB()) {
```
Comment (epiercehoffman) [inline src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:289]: I think the changes you made to the way CPX events are loaded in means that they are never interchromosomal by this check. Also, I am not sure that this will reliably identify insertion-type events, because if there is a DEL at the sink then POS != END - is that the desired behavior?

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java [hunk 3]
```diff
@@ -222,12 +274,31 @@ private static int getLength(final SVCallRecord record, final int missingInserti
...
+            // Insertion types use the sum of cpx intervals, or the "missing" length if there are none
+            if (record.getComplexEventIntervals().isEmpty()) {
+                return missingInsertionLength;
```
Comment (epiercehoffman) [inline src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:292]: Codecov is alerting me that you need a unit test for this case - although, shouldn't this throw an error if CPX_INTERVALS is empty for a CPX SV?

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java [hunk 4]
```diff
@@ -182,35 +192,77 @@ private static boolean clusterTogetherWithParams(final SVCallRecord a, final SVC
...
+        final List<SVCallRecord.ComplexEventInterval> intervalsB = b.getComplexEventIntervals();
+        if (intervalsA.size() != intervalsB.size()) {
+            return false;
+        }
+        final Iterator<SVCallRecord.ComplexEventInterval> iterA = intervalsA.iterator();
+        final Iterator<SVCallRecord.ComplexEventInterval> iterB = intervalsB.iterator();
+        for (int i = 0; i < intervalsA.size(); i++) {
+            final SVCallRecord.ComplexEventInterval cpxIntervalA = iterA.next();
+            final SVCallRecord.ComplexEventInterval cpxIintervalB = iterB.next();
+            if (cpxIntervalA.getIntervalType() != cpxIintervalB.getIntervalType()) {
```
Comment (epiercehoffman) [inline src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:212]: I'm wondering if you should also check the CPX_TYPE? Although I think this is sufficient, because even though you could technically get a dupINV and a dDUP with the same CPX_INTERVALS, they would have different CHROM and POS

File: src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java [hunk 5]
```diff
@@ -400,15 +408,191 @@ public void testClusterTogetherCNVs() {
...
+                {"chr1", 1000, "chr1", 2000,
+                        GATKSVVCFConstants.ComplexVariantSubtype.delINV,
+                        Arrays.asList(
+                                new SVCallRecord.ComplexEventInterval(GATKSVVCFConstants.StructuralVariantAnnotationType.DEL, new SimpleInterval("chr1", 800, 1100)),
+                                new SVCallRecord.ComplexEventInterval(GATKSVVCFConstants.StructuralVariantAnnotationType.INV, new SimpleInterval("chr1", 1600, 1900))
+                        ),
+                        false
+                },
+                // second cpx interval type is different
+                {"chr1", 1000, "chr1", 2000,
```
Comment (epiercehoffman) [inline src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java:477]: I think you need to update the coordinates to match the intervals to make sure this is testing what you want it to, because the overall coordinates get checked before the CPX_INTERVALS

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java [hunk 6]
```diff
@@ -222,12 +274,31 @@ private static int getLength(final SVCallRecord record, final int missingInserti
...
         } else {
             // TODO lengths less than 1 shouldn't be valid
             return Math.max(record.getLength() == null ? 1 : record.getLength(), 1);
         }
     }
 
+    /**
+     * Returns length to be used for complex variant matching
+     */
+    private static int getLengthComplex(final SVCallRecord record, final int missingInsertionLength) {
```
Comment (epiercehoffman) [inline src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:288]: It would be good to explicitly unit-test this instead of just testing the methods that call it

File: src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java [hunk 7]
```diff
@@ -400,15 +408,191 @@ public void testClusterTogetherCNVs() {
...
+                },
+                // not match by coordinates
+                {"chr1", 2000, "chr1", 3000,
+                        GATKSVVCFConstants.ComplexVariantSubtype.dDUP,
+                        Arrays.asList(
+                                new SVCallRecord.ComplexEventInterval(GATKSVVCFConstants.StructuralVariantAnnotationType.DUP, new SimpleInterval("chr1", 6200, 7990))
+                        ),
+                        false
+                },
+                // not match by subtype
```
Comment (epiercehoffman) [inline src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java:538]: It would be good to update this to look like a real dDUP_iDEL for a more accurate test (END coordinate, adding DEL interval to CPX_INTERVALS)

**Review by epiercehoffman — CHANGES_REQUESTED** [review]
Thanks for working on this - complex SVs make for complex code. My primary hesitation is related to the way length is calculated for CPX events - let me know what you think. There are some other smaller comments/questions as well, and you'll need to rebase as I anticipate conflicts with my recent CTX PR.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecordUtils.java [hunk 8]
```diff
@@ -34,34 +34,17 @@ public final class SVCallRecordUtils {
...
-        } else {
-            end = record.getPositionB();
-        }
+        final int end;
         final Integer end2;
         final String chr2;
         if (type == GATKSVVCFConstants.StructuralVariantAnnotationType.BND
-            || type == GATKSVVCFConstants.StructuralVariantAnnotationType.CTX) {
+                || type == GATKSVVCFConstants.StructuralVariantAnnotationType.CTX) {
+            end = record.getPositionA();
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecordUtils.java:43]: Made a TODO note. Would it also be possible to use the CPX intervals?

File: src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecord.java [hunk 9]
```diff
@@ -380,7 +399,73 @@ public Double getLog10PError() {
...
+    public List<ComplexEventInterval> getComplexEventIntervals() {
+        return cpxIntervals;
+    }
+
+    public static final class ComplexEventInterval implements Locatable {
+
+        private final GATKSVVCFConstants.StructuralVariantAnnotationType intervalType;
+        private final SimpleInterval interval;
+
+        public ComplexEventInterval(final String str) {
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecord.java:415]: This is a great point. This class is completely redundant with SVSegment, so I've moved that to its own class. However, I've made `ComplexEventInterval` a subclass with `encode()`/`decode()` methods. We could delete the new class altogether, but I think we should have this separate class since each is used in different tools. This will make it easier to make independent changes later if we need to.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecordUtils.java [hunk 10]
```diff
@@ -331,6 +321,7 @@ public static SVCallRecord create(final VariantContext variant, boolean keepVari
         final GATKSVVCFConstants.StructuralVariantAnnotationType type = inferStructuralVariantType(variant);
         final GATKSVVCFConstants.ComplexVariantSubtype cpxSubtype =
                 type == GATKSVVCFConstants.StructuralVariantAnnotationType.CPX ? getComplexSubtype(variant) : null;
+        final List<SVCallRecord.ComplexEventInterval> cpxIntervals = parseComplexIntervals(variant.getAttributeAsStringList(GATKSVVCFConstants.CPX_INTERVALS, "."));
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecordUtils.java:324]: Good catch, a better choice would be "null", although I think in either case it would lead to an error getting thrown. Looking at this had led me to discover that there were some problems with the unit tests (CPX type and intervals were not being checked), which I fixed, and I've added a dictionary check to the complex interval parsing.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java [hunk 11]
```diff
@@ -182,35 +192,77 @@ private static boolean clusterTogetherWithParams(final SVCallRecord a, final SVC
...
+            final SVCallRecord.ComplexEventInterval cpxIntervalA = iterA.next();
+            final SVCallRecord.ComplexEventInterval cpxIintervalB = iterB.next();
+            if (cpxIntervalA.getIntervalType() != cpxIintervalB.getIntervalType()) {
+                return false;
+            }
+            final SimpleInterval intervalA = cpxIntervalA.getInterval();
+            final SimpleInterval intervalB = cpxIintervalB.getInterval();
+            if (!(IntervalUtils.isReciprocalOverlap(intervalA, intervalB, overlapThreshold)
+                    && testSizeSimilarity(intervalA.getLengthOnReference(), intervalB.getLengthOnReference(), sizeSimilarityThreshold)
+                    && testBreakendProximity(new SimpleInterval(intervalA.getContig(), intervalA.getStart(), intervalA.getStart()),
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:219]: getPositionAInterval() is an SVRecord method, but intervalA is SimpleInterval

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java [hunk 12]
```diff
@@ -182,35 +192,77 @@ private static boolean clusterTogetherWithParams(final SVCallRecord a, final SVC
...
                                                  final SAMSequenceDictionary dictionary) {
-        final SimpleInterval intervalA1 = a.getPositionAInterval().expandWithinContig(window, dictionary);
-        final SimpleInterval intervalA2 = a.getPositionBInterval().expandWithinContig(window, dictionary);
-        if (intervalA1 == null) {
-            logger.warn("Invalid start position " + a.getPositionA() + " in record " + a.getId() +
+        return testBreakendProximity(a.getPositionAInterval(), a.getPositionBInterval(),
+                b.getPositionAInterval(), b.getPositionBInterval(), window, dictionary);
+    }
+
+    private static boolean testBreakendProximity(final SimpleInterval intervalA1, final SimpleInterval intervalA2,
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:250]: Do you mean the inputs in the call? Yes could have done it that way and avoided defining the overload with SVCallRecord inputs, but I think it's more readable this way.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecordUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (epiercehoffman) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/SVCallRecordUtils.java:43]: Just looked back at my conversation with Xuefang about this - for annotation we discussed that we should check CPX_INTERVALS for complex CTX, but in her CTX_INV example only the INV was in CPX_INTERVALS. For the CTX we used the other positions. If the INV is on the first chromosome, then the CTX breakpoints on that chromosome are defined by CHROM:POS and CHROM:END. But if the INV is on the second chromosome, then the CTX breakpoints are not fully defined by CHR2:END2 - there would need to be a CHR2:POS2.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java (hunk 1 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:294]: I agree, this check is probably unnecessary since we already test overlap of the individual complex intervals. I've removed this method entirely in favor of using the same check as INS types. If the variant does not have `SVLEN` assigned then `INSERTION_ASSUMED_LENGTH_FOR_OVERLAP` (50) is assumed, and the interval [POS, POS+SVLEN) is used for the overlap check.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java (hunk 2 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:289]: Removed this method

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java (hunk 3 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:292]: Removed this method. I didn't want to make any assumptions about whether CPX_INTERVALS is present, but I do think we need to draw up a spec for them.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java (hunk 4 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:212]: Right, `typesMatch()` looks at the subtypes.

File: src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java (hunk 5 above)
Comment (mwalker174) [reply src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java:477]: Not sure I'm following - this record should match `cpx1` on all checks, including the POS/END overlap, except for the second CPX interval type. Or is there a problem that the POS/END don't match any CPX interval?

File: src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java (hunk 6 above)
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java:288]: Agreed but it's deleted now

File: src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java (hunk 7 above)
Comment (mwalker174) [reply src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java:538]: I agree that it would be better to not have bogus records, but I think for testing purposes this should be left the way it is. If the records differ in any other way than the subtype, we don't know for sure if `areClusterable()` is returning  false due to the intended reason.

File: src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java (hunk 5 above)
Comment (epiercehoffman) [reply src/test/java/org/broadinstitute/hellbender/tools/sv/cluster/SVClusterEngineTest.java:493]: I think I got turned around, this makes sense - you are forcing the coordinates to match cpx1 to make sure that the intervals get checked

**Review by epiercehoffman — APPROVED** [review]
Thanks for making those changes!

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/GATKSVVCFConstants.java [hunk 13]
```diff
@@ -121,6 +123,13 @@ public enum ComplexVariantSubtype {
             entry("CTX_INV", ComplexVariantSubtype.CTX_INV)
     ));
 
+    // complex subtypes expected to have POS=END
+    public static final Set<ComplexVariantSubtype> COMPLEX_POINT_SUBTYPES =
+            Collections.set(
+                    GATKSVVCFConstants.ComplexVariantSubtype.dDUP,
+                    GATKSVVCFConstants.ComplexVariantSubtype.dDUP_iDEL
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/GATKSVVCFConstants.java:130]: Whoops. I was considering this but didn't end up using it
